DAGE-73: MtebWriter corpus bug

nicolo-rinaldi opened the pull request:
DAGE-73: [Jira ticket](https://sease.atlassian.net/browse/DAGE-73)

The bug was: MtebWriter produces corpus.json with only the documents inside the datastore -> should be all

Now this problem is addressed in the `main.py` script. We are creating a new instance of the datastore with autosave DISABLED, and we are adding to that all the documents from the search engine. In this way, the corpus file is built as expected. To fetch all documents from the search engine, a new method has been implemented in the BaseSearchEngine class: `.fetch_all()`. This calls the method  `fetch_for_generation()` multiple times. This choice has been made to avoid duplication of code.

Another slight change has been made: now the default output directory for embeddings is `resources/embeddings/` instead of `output/embeddings/`.

nicolo-rinaldi committed:
[DAGE-73] bug fixed by adding all doc to datastore after saving it

Needed changes to fix this bug:
- all search engines classes now have a new method: `.fetch_all()` that relies on `.fetch_for_query_generation()`, who has been slightly change with the start param. this choice has been made to avoid duplication of code.
- Found another bug related to solr-init.sh file: indexing with vector with too high precision is not working -> fixed by setting the embedding writer to write to file at a fixed precision (that can be changed in function param, now fixed at 12). FYI, Solr DenseVectorField seems to stop at 9 decimal numbers (rounding looking at the 10th)
- fixed some typos in mteb config file

This is mostly a temporary solution working without touching tons of code (just main script and search engines, that would have been touched anyway). Seems to work as expected

Executed:
- ruff
- mypy
- test (need still to update them by adding test for `.fetch_all()` method)

# Conflicts:
#	rre-tools/dataset-generator/src/dataset_generator/search_engine/search_engine_base.py

# Conflicts:
#	rre-tools/embedding-model-evaluator/config.yaml
#	rre-tools/embedding-model-evaluator/src/embedding_model_evaluator/config.py

nicolo-rinaldi committed:
[DAGE-73] added tests

nicolo-rinaldi committed:
[DAGE-73] Updated version with new DataStore autosave feature

nicolo-rinaldi committed:
[DAGE-73] added `uv sync --group dev` to README.md

Intrinsical-AI commented on rre-tools/dataset-generator/src/dataset_generator/main.py:
```diff
@@ -149,5 +147,17 @@ def main() -> None:
+        all_doc: List[Document] = search_engine.fetch_all(doc_fields=config.doc_fields)
+        for doc in all_doc:
+            data_store.add_document(doc)
+        data_store = new_data_store
```
**[BUG]**

Why adding to old database, and then assign the old database to the new database?

I think with the current implementation we're always writing an empty data_store (?)


Maybe would be simpler just add to the new DS and then write the new DS in that case; data_store else.
```
if config.output_format == "mteb":
        new_data_store: DataStore = DataStore()
        all_doc: List[Document] = search_engine.fetch_all(doc_fields=config.doc_fields)
        for doc in all_doc:
            new_data_store.add_document(doc)
        writer.write(output_destination, new_data_store)
else:
        writer.write(output_destination, data_store)
```

For example?

Intrinsical-AI replied:
If we would wanted to keep a single datastore, this might be also possible:

```
if config.output_format == "mteb":
    ids = {d.id for d in data_store.iter_documents()}
    for doc in search_engine.fetch_all(doc_fields=config.doc_fields):
        if doc.id not in ids:
            data_store.add_document(doc)
    # opcional: data_store.save() - caution to configure correcly autosave if implemented
writer.write(output_destination, data_store)
```
Anw: I think is more simpler and robust robust use 2 distinct DS instances by now and let it be till we polish the main pipeline / auxiliars [my take]

nicolo-rinaldi replied:
I will resolve with the first solution. That was a lint complaint and I changed too fast

Intrinsical-AI commented on rre-tools/dataset-generator/src/dataset_generator/search_engine/search_engine_base.py:
```diff
@@ -6,12 +6,32 @@
+        all_doc: List[Document] = []
+        start = 0
+        while True:
+            all_doc.extend(self.fetch_for_query_generation(documents_filter=None,
```
**[Robustness]**

Maybe checking

```
batch = self.fetch_for_query_generation(..)
if not batch: # avoid adding None, [], ...
        break
all_docs.extend(batch)
start += len(batch) # small improvement: len(batch) can be < DOC_NUMBER_EACH_FETCH
```

Intrinsical-AI replied:
Also: I realize this induces some conflicts with the [Vespa-PR](https://github.com/SeaseLtd/rated-ranking-evaluator/pull/228) (again). I've just spend some more time solving conflicts and testing. Since this can be implemented outside the base engine; are you OK moving this code to either an auxiliar file, to the main script, or somewhere else to avoid inducing conflicts? Would that be possible?

nicolo-rinaldi replied:
Changed

Intrinsical-AI replied:
@nicolo-rinaldi thanks man!

Intrinsical-AI commented on rre-tools/embedding-model-evaluator/src/embedding_model_evaluator/writers/embedding_writer.py:
```diff
@@ -14,12 +14,15 @@
+                vector = np.round(vector, precision).tolist()
+            vector = [float(f"{x:.{precision}e}") for x in vector]
+            # format each float to a fixed precision string, then convert back to float
             jsonl.write({"id": _id, "vector": vector})
```
**[Suggestion]**

Cast to common type, check, and write to json avoiding float re-conversion (I belive JSON should handle as proposed; list primitive implements `_str_`)
```
arr = np.asarray(vector, dtype=np.float32)
if not np.isnan(arr) and np.isfinite(arr).all(): # Check: no n.a.ns // no infinite vals
 jsonl.write({"id": _id, "vector": arr.tolist()})
else:
 raise ...
```

nicolo-rinaldi replied:
I already tried this. It doesn't work

Intrinsical-AI replied:
The problem I see is that currently we're rounding 2 times. The 2nd time, when casting string -> float we pass from "rounded-but-full-prec" to "rounded-and-rerounded-again".

With expected vectors this should work, and give distinct results than current impl. (checked on Colab with json writting)
```
..
            if isinstance(vector, np.ndarray):
                vector = vector.round(decimals=precision).tolist()
            else:
                vector = [round(float(x), precision) for x in vector]
```

> If we're getting an unexpected error when casting or something like that might be corrupt data / input format missmatch or loading (None, [], np.nan, "", "'"...)

PS: the method sign currently is:
`items: Iterable[tuple[str, np.ndarray | list[float]]]`

I think it could be:
` Iterable[tuple[str, np.ndarray | Sequence[float]]] ` 

Sequence accepts wider range of types as tuples, torch.Tensor/tf.Tensor, pandas.Serie..

nicolo-rinaldi replied:
Ok, I'll try this and change it if it works

nicolo-rinaldi replied:
At the end, I rolled back. The bug was no more there

nicolo-rinaldi replied:
Bug actually was there, I changed the solr-init to solve only for solr

Intrinsical-AI commented on rre-tools/README.md:
```diff
@@ -34,8 +34,8 @@ cd rre-tools
-# install optional dev dependencies such as mypy/ruff
-uv sync --extra dev
+# install development dependencies as well (e.g., mypy and ruff)
+uv sync --group dev
```
**[Note]**

I think this are included in uv default profile (used with the plain uv sync). If so this is not needed but anyway I like it making it being explicit.

Intrinsical-AI commented on rre-tools/dataset-generator/tests/unit/test_elasticsearch_search_engine.py:
```diff
@@ -64,6 +65,27 @@ def test_elasticsearch_search_engine_fetch_for_evaluation__expects__result_retur
                                                 doc_fields=elasticsearch_config.doc_fields)
     assert result[0] == Document(**mock_dict)
 
+def test_elasticsearch_engine_fetch_all__expects__results_returned(monkeypatch, elasticsearch_config, mock_doc, mock_dict):
```
Nice adding tests

Intrinsical-AI commented on rre-tools/dataset-generator/tests/unit/test_opensearch_engine.py:
```diff
@@ -18,6 +19,16 @@ def opensearch_config(resource_folder):
     return Config.load(resource_folder / "good_config_opensearch.yaml")
 
+@pytest.fixture
+def expected_doc():
```
good mock

Intrinsical-AI commented on rre-tools/embedding-model-evaluator/src/embedding_model_evaluator/config.py:
```diff
@@ -18,11 +18,11 @@ class Config(BaseModel):
     embeddings_dest: Optional[Path] = Field(
         None,
-        description="Path to save mteb embeddings, by default saved in <output/embeddings> folder.",
+        description="Path to save mteb embeddings, by default saved in <resource/embeddings> folder.",
```
Not related // conflicts with DAGE-65 but ok let's keep

Intrinsical-AI commented on rre-tools/embedding-model-evaluator/config.yaml:
```diff
@@ -19,9 +19,9 @@ candidates_path: "resources/data/candidates.jsonl"
 output_dest: "resources"
 
-# (Optional) Path to write mteb document and query embeddings, if not given it will be written to output/embeddings dir
+# (Optional) Path to write mteb document and query embeddings, if not given it will be written to resources/embeddings dir
```
Not related // conflicts with DAGE-65

In general for this things, I'd make a general refactor PR only for comments / namings / docs; updating and cleaning.  Ok to keep as it is this time

Intrinsical-AI reviewed (commented):
One potential bug / Some minor points to agree

nicolo-rinaldi committed:
[DAGE-73] typo in main.py datastore variable

nicolo-rinaldi committed:
[DAGE-73] added robustness to fetch_all

nicolo-rinaldi committed:
[DAGE-73] rolled back to `_write_embeddings_jsonl` old version in EmbeddingWriter

nicolo-rinaldi committed:
[DAGE-73] 2 big fixes

Bugs:
1. fixed the indexing solr bug directly in the solr-init.sh file, rounding to the 12th decimal number (this choice has been made due to Solr errors with too precise decimals, while still maintaining the numpy default precision)
2. fixed bug in accessing the cache in embedding_writer.py: the param name was not allowed. The param task_name is enough since the process is a hashing step and we don't need to specify if a text comes from the specific split of the dataset

# Conflicts:
#	rre-tools/docker-services/solr-init/solr-init.sh
#	rre-tools/embedding-model-evaluator/src/embedding_model_evaluator/writers/embedding_writer.py

nicolo-rinaldi commented:
Tests are now failing. Wait to review until I re-design the ones that are not correct now

nicolo-rinaldi committed:
[DAGE-73] reverted back max_tokens to be ignored by ruff

nicolo-rinaldi committed:
[DAGE-73] workaround modified to avoid changing the datastore persistency

Intrinsical-AI reviewed (approved):
I had this pending👍

nicolo-rinaldi committed:
[DAGE-73] fixed description bug

nicolo-rinaldi committed:
[DAGE-73] extended fetch_all() method to be used with vespa_search_engine.py

nicolo-rinaldi committed:
[DAGE-73] uniformed config param in M1 and M2

nicolo-rinaldi committed:
[DAGE-73] centralized function setup_logging

nseidan commented on rre-tools/dataset-generator/src/dataset_generator/search_engine/search_engine_base.py:
```diff
@@ -6,12 +6,32 @@
+        # Now this is relying on fetch_for_query_generation to avoid duplicate code. Might be changed in the future
+        all_docs: List[Document] = []
+        start: int = 0
+        while True:
```
while true - dangerous statement -> leads to stackoverflow

dantuzi replied:
You can run the query the first time, get the total number of results (on Solr, you have the `numFound` value, in ElasticSearch/OpenSearch you have `hits.total.value`), and iterate while the start + rows < total number of documents.

Another alternative is to directly add an abstract generic `search` method in the BaseSearchEngine class that is supposed to implement the paginated search. This is extended in all SearchEngine clients, and each client implements the pagination using the specific SearchEngine API.
For example, in Solr we can only use the `start` and `rows` parameters but in ElasticSearch we can use the `search_after` parameter (https://www.elastic.co/docs/reference/elasticsearch/rest-apis/paginate-search-results).
That being said, it's probably easier to refactor the search engine classes in the future

nseidan commented on rre-tools/dataset-generator/src/dataset_generator/main.py:
```diff
@@ -139,15 +138,33 @@ def main() -> None:
+    #  work on a better solution, instead of overwriting the corpus.json file, and maybe modify the MtebWriter with the
+    #  fetch from the search engine
+    if config.output_format == "mteb":
+        all_docs: List[Document] = search_engine.fetch_all(doc_fields=config.doc_fields)
```
in case of few hundred thousands  docs in the engine, we will hit the buffer limit by querying all docs. Pass a threshold there?

dantuzi replied:
Instead of collecting all documents and storing them in a file, we should explore how we can implement a kind of streaming.
A possibility can be to pass a callback function to the method `fetch_all` that is invoked for every new document read from the search engine.
The implementation of the callback appends a new line to the `corpus.jsonl` file

nicolo-rinaldi replied:
One by one is going to raise problems opening the file multiple times, I'm going to do a batch version of it

nicolo-rinaldi replied:
done

dantuzi commented on rre-tools/dataset-generator/src/dataset_generator/main.py:
```diff
@@ -139,15 +138,33 @@ def main() -> None:
+
+        # copy pasted from MtebWriter
+        corpus_path = Path(output_destination) / "corpus.jsonl"
+        with corpus_path.open("w", encoding="utf-8") as file:
+            for doc in all_docs:
+                doc_id = str(doc.id)
+                fields = doc.fields
+                title = _to_string(fields.get("title"))
+                text = _to_string(fields.get("description"))
+
+                row = {"id": doc_id, "title": title, "text": text}
+                file.write(json.dumps(row, ensure_ascii=False) + "\n")
```
The fields `title` and `description` are hardcoded

nicolo-rinaldi replied:
fixed

nicolo-rinaldi committed:
[DAGE-73] fixed hardcoded description and coded the streamed version of fetch_all(), that uses a prefixed batch of 100

nicolo-rinaldi committed:
[DAGE-73] added _get_total_hits to do the streamed version of fetch_all for all search engines but vespa. Also fixed tests

nicolo-rinaldi committed:
[DAGE-73] vespa should be fixed, I need to update the tests since fetch_all related ones are missing

nicolo-rinaldi committed:
[DAGE-73] fixed vespa tests

nicolo-rinaldi commented:
@dantuzi @nseidan I fixed what you asked. Now there are a lot more files changed, but I hope the fixes are good

nseidan commented on rre-tools/dataset-generator/tests/unit/mocks/elasticsearch.py:
```diff
@@ -2,14 +2,18 @@
             "hits": {
-                "hits": self._json_data
+                "hits": self._json_data,
+                "total": {
```
maybe use`total_hits` as you used in the other code

nicolo-rinaldi replied:
Sorry, I'm not getting this comment. This should be the same of the other code

nseidan replied:
I meant renaming it. I am seeing sometimes total_hits

nicolo-rinaldi replied:
still not getting you. I should have written `total_hits` everywhere I'm using this (e.g., mocks.vespa.py, mocks.solr.py, mocks.elasticsearch.py, mocks.opensearch.py)

nseidan commented on rre-tools/dataset-generator/src/dataset_generator/search_engine/search_engine_base.py:
```diff
@@ -2,16 +2,42 @@
+        # Now this is relying on fetch_for_query_generation to avoid duplicate code. Might be changed in the future
+        start: int = - DOC_NUMBER_EACH_FETCH
+        total_hits: int = self._get_total_hits(self._fetch_all_payload)
+        while start + DOC_NUMBER_EACH_FETCH < total_hits:
```
-> `start` (line#29)   is a negative number and passing it to directly `fetch_for_query_generation()`  so it will break. 
 -> not checking the batch if it is empty , in this case, it will be in an infinite loop. 
-> while statement does not handle the last page when `start + DOC_NUMBER_EACH_FETCH == total_hits`
How about starting with a simple while statement:
```
  start = 0
  page_size = DOC_NUMBER_EACH_FETCH
  total_hits: int = self._get_total_hits(...)
  while start < total_hits:
       code ...
```

Another case is about total_hits if it is changed (e.g. search engine collection/index is updated with more data, total_hits becomes stale). Maybe for now we don't consider this case

nicolo-rinaldi replied:
Tried to fix

nseidan commented on rre-tools/dataset-generator/src/dataset_generator/search_engine/solr_search_engine.py:
```diff
@@ -24,10 +24,37 @@ def __init__(self, endpoint: HttpUrl):
+        log.debug(f"Payload: {payload}")
+
+        try:
+            response = requests.post(search_url, headers=self.HEADERS, json=payload)
```
why POST here for solr? Normally /select is used with GET ?

nicolo-rinaldi replied:
We used POST everywhere. If we change this, we need to do it also for all other search engines

nseidan commented on rre-tools/dataset-generator/src/dataset_generator/search_engine/vespa_search_engine.py:
```diff
@@ -42,6 +42,31 @@ def __init__(self, endpoint: HttpUrl):
+    @property
+    def _fetch_all_payload(self) -> Dict[str, Any]:
+        return {
+            'yql': f"select * from {self.schema} where true"
```
why to use `where true` it doesn't reflect anything?

nicolo-rinaldi replied:
Removed

nicolo-rinaldi committed:
[DAGE-73] addressing Naz comments

nseidan commented on rre-tools/dataset-generator/src/dataset_generator/search_engine/search_engine_base.py:
```diff
@@ -2,16 +2,45 @@
+                doc_fields=doc_fields,
+                start=start
+                )
+            yield batch
```
before yield, check if batch is empty/None, if so then `break` the loop. Otherwise, it will be in the infinite loop.

nicolo-rinaldi replied:
done

nicolo-rinaldi replied:
Done

nicolo-rinaldi replied:
Done

nicolo-rinaldi committed:
[DAGE-73] added break in while loop

dantuzi commented on rre-tools/dataset-generator/src/dataset_generator/search_engine/elasticsearch_search_engine.py:
```diff
@@ -37,12 +57,13 @@ def fetch_for_query_generation(self,
         """
         # Build base query
-        query: Dict[str, Any] = {"match_all": {}}
+        query: Dict[str, Any] = self._fetch_all_payload
```
The function _fetch_all_payload should be invoked.

```
query: Dict[str, Any] = self._fetch_all_payload()
```

nicolo-rinaldi replied:
since it's a property, it should not be invoked

dantuzi commented on rre-tools/dataset-generator/src/dataset_generator/search_engine/opensearch_engine.py:
```diff
@@ -42,20 +61,20 @@ def fetch_for_query_generation(self,
-            query = {
-                "match_all": {}
-            }
+            query = self._fetch_all_payload
```
Same as before

dantuzi commented on rre-tools/dataset-generator/src/dataset_generator/search_engine/search_engine_base.py:
```diff
@@ -2,16 +2,47 @@
+                )
+            if not batch:
+                break
+            yield batch
```
It would be simpler if we yield a single document, not a batch.
In other words, I would iterate for each document in the batch and yield each document.

nicolo-rinaldi replied:
fixed

dantuzi commented on rre-tools/dataset-generator/src/dataset_generator/main.py:
```diff
@@ -139,15 +138,37 @@ def main() -> None:
+        corpus_path.unlink(missing_ok=True)
+        for docs_batch in search_engine.fetch_all(doc_fields=config.doc_fields):
+            with corpus_path.open("a", encoding="utf-8") as file:
+                for doc in docs_batch:
```
Why do you open the file for writing the single batch?
How about the following solution? (I'm considering that the fetch_all iterates on documents, not batches)

```
with corpus_path.open("a", encoding="utf-8") as file:
   for docs in search_engine.fetch_all(doc_fields=config.doc_fields):
      doc_id = str(doc.id)
      fields = doc.fields
      title = _to_string(fields.get("title"))
      text = " ".join(
      _to_string(value)
      for key, value in fields.items()
      if key != "title"
      )
      
      row = {"id": doc_id, "title": title, "text": text}
      file.write(json.dumps(row, ensure_ascii=False) + "\n")
```

nicolo-rinaldi replied:
fixed

nicolo-rinaldi committed:
[DAGE-73] addressing Daniele's comments
